Output implies that Citations are lower (and Warnings higher) than expected for Asian, Other, and White drivers. Conversely, Citations are higher for Black drivers, and highest for Hispanic drivers.

In [25]:
import sys
sys.path.append(r"C:/Users/mikha/Dropbox/mikhael_misc/Projects/My-Package")

import pandas as pd
import numpy as np
from myfunctions import clean_path

from scipy.stats import chi2_contingency

from IPython.display import display, HTML

df = pd.read_csv(filepath_or_buffer=clean_path(r"C:/Users/mikha/Dropbox/mikhael_misc/Projects/Policing Thesis/Modified Dataset - 2021.csv")).set_index('Stop ID')

C:\Users\mikha\Anaconda\lib\site-packages\IPython\core\interactiveshell.py:3172: DtypeWarning: Columns (18,19,20,21,22,23,32,56) have mixed types.Specify dtype option on import or set low_memory=False.
  has_raised = await self.run_ast_nodes(code_ast.body, cell_name,


In [26]:
df['Fatal - Sparse Filled'].sum()

434.0

In [27]:
accident_df = df[df['Accident - Sparse Filled']==1]
accident_df.name='Accident DF'

fatality_df = df[df['Fatal - Sparse Filled']==1]
fatality_df.name='Fatality DF'

personal_injury_df = df[df['Personal Injury - Sparse Filled']==1]
personal_injury_df.name='Personal Injury DF'


## Chi-Square Test

In [32]:
def chi_square_contingency_test(dataframe:pd.DataFrame) -> pd.DataFrame:

    contingency_table = pd.crosstab([dataframe['Race']], [dataframe['Violation Type']])


    chi2_result = chi2_contingency(contingency_table)

    print(f"Test statistic == {chi2_result[0]}")
    print(f"P-Value == {'{:.40f}'.format(chi2_result[1])}")
    print(f"Degrees of Freedom == {chi2_result[2]}")


    print("Contingency table == ")
    display(contingency_table)

    expected_frequencies = pd.DataFrame(chi2_result[3])
    expected_frequencies.columns = contingency_table.columns
    expected_frequencies.index = contingency_table.index
    print(f"Expected Frequencies == ")
    display(expected_frequencies.round(0))

    print(f"Difference between Actual and Expected Frequencies ({dataframe.name}) (Actual - Expected) == \n (Negative => ")
    display(contingency_table-chi2_result[3].round(0))

## Accidents

In [33]:
chi_square_contingency_test(accident_df)

Test statistic == 394.89335625429
P-Value == 0.0000000000000000000000000000000000000000
Degrees of Freedom == 15
Contingency table == 


Violation Type   Citation  ESERO  SERO  Warning
Race                                           
ASIAN                2607     25     0      374
BLACK               13298     82     4     1366
HISPANIC            14982     47     3      935
NATIVE AMERICAN       100      2     0       14
OTHER                2391     14     3      358
WHITE               17184     82     0     2050

Expected Frequencies == 


Violation Type   Citation  ESERO  SERO  Warning
Race                                           
ASIAN              2718.0   14.0   1.0    274.0
BLACK             13336.0   66.0   3.0   1344.0
HISPANIC          14437.0   72.0   3.0   1455.0
NATIVE AMERICAN     105.0    1.0   0.0     11.0
OTHER              2501.0   12.0   0.0    252.0
WHITE             17465.0   87.0   3.0   1761.0

Difference between Actual and Expected Frequencies (Accident DF) (Actual - Expected) == 
 (Negative => 


Violation Type   Citation  ESERO  SERO  Warning
Race                                           
ASIAN              -111.0   11.0  -1.0    100.0
BLACK               -38.0   16.0   1.0     22.0
HISPANIC            545.0  -25.0   0.0   -520.0
NATIVE AMERICAN      -5.0    1.0   0.0      3.0
OTHER              -110.0    2.0   3.0    106.0
WHITE              -281.0   -5.0  -3.0    289.0

## Fatalities

Note that there have been few fatalities (434), so it's hard to draw any conclusions from this table.

In [34]:
chi_square_contingency_test(fatality_df)

Test statistic == 20.240673991875365
P-Value == 0.0270567605498549523801443683623801916838
Degrees of Freedom == 10
Contingency table == 


Violation Type   Citation  ESERO  Warning
Race                                     
ASIAN                  21      0        4
BLACK                  97      3       37
HISPANIC               62      2       32
NATIVE AMERICAN         0      0        1
OTHER                   8      0       12
WHITE                  86      3       66

Expected Frequencies == 


Violation Type   Citation  ESERO  Warning
Race                                     
ASIAN                16.0    0.0      9.0
BLACK                86.0    3.0     48.0
HISPANIC             61.0    2.0     34.0
NATIVE AMERICAN       1.0    0.0      0.0
OTHER                13.0    0.0      7.0
WHITE                98.0    3.0     54.0

Difference between Actual and Expected Frequencies (Fatality DF) (Actual - Expected) == 
 (Negative => 


Violation Type   Citation  ESERO  Warning
Race                                     
ASIAN                 5.0    0.0     -5.0
BLACK                11.0    0.0    -11.0
HISPANIC              1.0    0.0     -2.0
NATIVE AMERICAN      -1.0    0.0      1.0
OTHER                -5.0    0.0      5.0
WHITE               -12.0    0.0     12.0

## Personal Injury

In [35]:
chi_square_contingency_test(personal_injury_df)

Test statistic == 67.71710453179797
P-Value == 0.0000000001220863624762970545741272547513
Degrees of Freedom == 10
Contingency table == 


Violation Type   Citation  ESERO  Warning
Race                                     
ASIAN                1060      1      111
BLACK                5004      9      356
HISPANIC             5333      5      309
NATIVE AMERICAN        41      0        1
OTHER                 917      0      111
WHITE                6888      5      586

Expected Frequencies == 


Violation Type   Citation  ESERO  Warning
Race                                     
ASIAN              1088.0    1.0     83.0
BLACK              4982.0    5.0    382.0
HISPANIC           5240.0    5.0    401.0
NATIVE AMERICAN      39.0    0.0      3.0
OTHER               954.0    1.0     73.0
WHITE              6940.0    7.0    532.0

Difference between Actual and Expected Frequencies (Personal Injury DF) (Actual - Expected) == 
 (Negative => 


Violation Type   Citation  ESERO  Warning
Race                                     
ASIAN               -28.0    0.0     28.0
BLACK                22.0    4.0    -26.0
HISPANIC             93.0    0.0    -92.0
NATIVE AMERICAN       2.0    0.0     -2.0
OTHER               -37.0   -1.0     38.0
WHITE               -52.0   -2.0     54.0